hivemind/moe/client/expert.py

```diff
@@ -52,18 +81,171 @@ def forward(self, *args, **kwargs):
+        if return_future:
+
+            async def _unpack(infos_future: MPFuture, dht: DHT):
+                p2p = await dht.replicate_p2p()
+                return [cls._spawn_experts(i, p2p) for i in await infos_future]
+
+            return cls.run_coroutine(_unpack(infos, dht), return_future)
+
+        return [cls.spawn_experts(exps, dht) for exps in infos]
+
+
+async def _backward_stream(uid: str, serialized_tensors: Iterable[runtime_pb2.Tensor], stub) -> List[torch.Tensor]:
+    split = (p for t in serialized_tensors for p in split_for_streaming(t, DEFAULT_MAX_MSG_SIZE // 2))
+
+    grad_inputs = await stub.rpc_backward_stream(
+        amap_in_executor(
+            lambda tensor: runtime_pb2.ExpertRequest(uid=uid, tensors=[tensor]),
+            iter_as_aiter(split),
+        ),
+    )
+
+    return await gather_from_streaming(grad_inputs, lambda r: r.tensors, deserialize_torch_tensor)
+
+
+async def _backward_unary(uid: str, serialized_tensors: Iterable[runtime_pb2.Tensor], stub) -> List[torch.Tensor]:
+    grad_inputs: runtime_pb2.ExpertResponse = await stub.rpc_backward(
+        runtime_pb2.ExpertRequest(uid=uid, tensors=list(serialized_tensors))
+    )
+    return [deserialize_torch_tensor(t) for t in grad_inputs.tensors]
+
+
+async def expert_backward(
+    uid: str, inputs_and_grads: Sequence[torch.Tensor], compressions: Iterable, stub
+) -> List[torch.Tensor]:
+    serialized_tensors = (
+        serialize_torch_tensor(tensor, compression) for tensor, compression in zip(inputs_and_grads, compressions)
+    )
+
+    size = 0
+    for t in inputs_and_grads:
+        size += t.element_size() * t.nelement()
+        if size >= DEFAULT_MAX_MSG_SIZE:
+            return await _backward_stream(uid, serialized_tensors, stub)
+    else:
+        return await _backward_unary(uid, serialized_tensors, stub)
+
+
+async def _forward_stream(uid: str, serialized_tensors: Iterable[runtime_pb2.Tensor], stub) -> List[torch.Tensor]:
+    split = (p for t in serialized_tensors for p in split_for_streaming(t, DEFAULT_MAX_MSG_SIZE // 2))
+
+    outputs = await stub.rpc_forward_stream(
+        amap_in_executor(
+            lambda tensor: runtime_pb2.ExpertRequest(uid=uid, tensors=[tensor]),
+            iter_as_aiter(split),
+        ),
+    )
+
+    return await gather_from_streaming(outputs, lambda r: r.tensors, deserialize_torch_tensor)
+
+
+async def _forward_unary(uid: str, serialized_tensors: Iterable[runtime_pb2.Tensor], stub) -> List[torch.Tensor]:
+    outputs: runtime_pb2.ExpertResponse = await stub.rpc_forward(
+        runtime_pb2.ExpertRequest(uid=uid, tensors=list(serialized_tensors))
+    )
+    return [deserialize_torch_tensor(t) for t in outputs.tensors]
+
+
+async def expert_forward(uid: str, inputs: Sequence[torch.Tensor], compressions: Iterable, stub) -> List[torch.Tensor]:
+    serialized_tensors = (
+        serialize_torch_tensor(tensor, compression) for tensor, compression in zip(inputs, compressions)
+    )
```

Comment:
By the way, why did you decide to serialize tensors inside this coroutine instead of doing it before the function call? The execution flow seems to be the same

Reply:
The thing here is that after the remote call of forward/backward we get `AsyncIterator` back, because tensors are being streamed. 

To make things the way you ask we have to materialize both inputs and outputs. We have to make a list of serialized tensors, pass them into coroutine, then after getting the results from net, we have to make a list of serialized tensors inside coroutine, and only then  deserialize them outside of async and create a list `torch.Tensors`.

The problem I see is materializing lists on every tiny step. On the other hand I can't say right now if it is more costly then serialization/deserialization itself.

You think this way is better than now? (we have to add +2 materializations of python list/tuple, instead of processing "on the fly")

By the way, I think that I did not quite what you asked: I just moved generator out of coroutine. But this does not mean that serialization happens outside. Generators are lazy, so serialization is still performed inside coroutine, only when we begin to iterate over parts of genrator. If i understand it the right way, serialization happens inside the executor which is hidden inside `amap_in_executor`